Expand state channels developer docs (#6869)
ethereum/ethereum-org-website: src/content/developers/docs/scaling/state-channels/index.md

The delay comes into play in single-user exits due to the possibility of fraudulent actions. For example, a channel participant may try to finalize the channel on Ethereum by submitting an older state update on-chain. 

As a countermeasure, state channels allow honest users to challenge invalid state updates by submitting the latest, valid state of the channel on-chain. State channels are designed such that newer, agreed-upon state updates trump older state updates.

Once a peer triggers the on-chain dispute-resolution system, the other party is required to respond within a time limit (called the challenge window). This allows users to challenge the exit transaction, especially if the other party is applying a stale update. 

Whatever the case may be, channel users always have strong finality guarantees: if the state transition in their possession was signed by all members and is the most recent update, then it is of equal finality with a regular on-chain transaction. They still have to challenge the other party on-chain, but the only possible outcome is finalizing the last valid state, which they hold. 

> #### How do state channels interact with Ethereum? {#how-do-state-channels-interact-with-ethereum}

Reviewer: I think this is supposed to be 3rd level header?
```suggestion
### How do state channels interact with Ethereum? {#how-do-state-channels-interact-with-ethereum}
```

incomplete: true
sidebarDepth: 3
---
State channels allow participants to securely transact off-chain while keeping interaction with Ethereum Mainnet at a minimum. Channel peers can conduct an arbitrary number of off-chain transactions while only submitting two on-chain transactions to open and close the channel. This allows for extremely high transaction throughput and results in lower costs for users. 

## Prerequisites {#prerequisites}
You should have read and understood our pages on [Ethereum scaling](/developers/docs/scaling/) and [layer 2](/layer-2/).

## What are channels? {#what-are-channels}

> Public blockchains, such as Ethereum, experience scalability issues due to their distributed architecture: on-chain transactions must be executed by all nodes, even if the said transactions concern a small number of users. Blockchain channels solve this problem by allowing users to keep interactions off the main blockchain.

Reviewer: ```suggestion
Public blockchains, such as Ethereum, face scalability challenges due to their distributed architecture: on-chain transactions must be executed by all nodes. Nodes have to be able to handle the volume of transactions in a block using modest hardware, imposing a limit on the transaction throughput to keep the network decentralized. Blockchain channels solve this problem by allowing users to interact off the main blockchain.
```

> Channels are simple bi-directional, peer-to-peer protocols that allow two parties to execute transactions off-chain. A channel functions as a fraud proof-based, multisignature state machine replicated by two P2P nodes. 

Reviewer: This is extremely jargon-heavy and probably doesn't help someone trying to learn about state-channels for the first time.
```suggestion
Channels are simple bi-directional, peer-to-peer protocols that allow two parties to make many transactions between themselves and then only post the final results to the blockchain. The channel uses cryptography to demonstrate that the summary data they generate is truly the result of a valid set of intermediate transactions. A ["multisig"](developers/docs/smart-contracts/#multisig) smart contract ensures the transactions are signed by the correct parties.
```

> Channel peers interact by proposing state updates and signing updates from others. Blockchain channels require all participants to sign state updates, hence their description as "multisignature state machines."

Reviewer: ```suggestion

```

With channels, state changes are executed and validated by interested parties, minimizing computation on Ethereum's execution layer. This decreases congestion on Ethereum and also increases transaction processing speeds for users. 

Each channel is managed by a [multisig smart contract](/developers/docs/smart-contracts/#multisig) running on Ethereum. To open a channel, participants deploy the channel contract on-chain and deposit funds into it. Both parties collectively sign a state update to initialize the channel's state, after which they can transact quickly and freely off-chain. 

> To close the channel, participants submit the last agreed-upon state of the channel on-chain. Afterward, the smart contract distributed the locked funds according to each participant's balance in the channel's final state. 

Reviewer 2: ```suggestion
To close the channel, participants submit the last agreed-upon state of the channel on-chain. Afterward, the smart contract distributes the locked funds according to each participant's balance in the channel's final state. 
```

Peer-to-peer channels are particularly useful for situations where some predefined participants wish to transact with high frequency without incurring visible overhead. Blockchain channels fall under two categories: **payment channels** and **state channels**. 

## Payment channels {#payment-channels}

A payment channel is best described as a "two-way ledger" collectively maintained by two users. The ledger's initial balance is the sum of deposits locked into the on-chain contract during the channel opening phase. 

Updates to the ledger's balance (i.e., the payment channel's state) require the approval of all parties in the channel. A channel update, signed by all channel participants, is considered finalized, much like a transaction on Ethereum. 

Payment channels were among the earliest scaling solutions designed to minimize expensive on-chain activity in simple user interactions (e.g., ETH transfers, atomic swaps, micropayments). Channel participants can conduct an unlimited amount of instant, feeless transactions between each other as long as the net sum of their transfers does not exceed the deposited tokens.

## State channels {#state-channels}

> Although useful in many ways, payment channels had limited applications. Beyond supporting off-chain payments, payment channels weren't useful for handling general state transition logic. State channels were created to solve this problem and make channel designs useful for scaling general-purpose computation. 

Reviewer: ```suggestion
Apart from supporting off-chain payments, payment channels have not proven useful for handling general state transition logic. State channels were created to solve this problem and make channels useful for scaling general-purpose computation. 
```

> State channels built on the idea of off-chain execution and extended it from simple state updates (e.g., token transfers) to arbitrary state transitions. State channels are designed to support general computation, making them more useful than ordinary payment channels. 

Reviewer: Effectively repeats the previous point.

```suggestion

```

> State channels still have a lot in common with payment channels. For example, users interact by exchanging cryptographically signed messages (transactions), which the other must also sign. If a proposed state update isn't signed by all participants, it cannot achieve validity. 

Reviewer: ```suggestion
State channels still have a lot in common with payment channels. For example, users interact by exchanging cryptographically signed messages (transactions), which the other must also sign. If a proposed state update isn't signed by all participants, it is not valid. 
```

> The major difference comes from the design of state channels. In addition to holding the ledger balance, the channel's state also references the current state of the contract's storage (i.e., values of contract variables). 

Reviewer: ```suggestion
However, in addition to holding the user's balances, the channel also tracks the current state of the contract's storage (i.e., values of contract variables). 
```

This makes it possible to execute a smart contract off-chain between two users. In this scenario, updates to the smart contract's internal state require only the approval of the peers who created the channel. 

While this solves the scalability problem described earlier, it has implications for security. On Ethereum, the validity of state transitions on Ethereum is enforced by the network's consensus protocol. This makes it impossible to propose an invalid update to a smart contract's state or alter smart contract execution. 

> State channels don't have the same security guarantees. To some extent, a state channel is a miniature version of Mainnet...with two nodes. With a limited set of participants enforcing rules, the possibility of malicious behavior (e.g., proposing invalid state updates) increases. To ensure security, state channels rely on a dispute arbitration system based on [fraud proofs](/glossary/#fraud-proof). 

Reviewer: ```suggestion
State channels don't have the same security guarantees. To some extent, a state channel is a miniature version of Mainnet. With a limited set of participants enforcing rules, the possibility of malicious behavior (e.g., proposing invalid state updates) increases. State channels derive their security from a dispute arbitration system based on [fraud proofs](/glossary/#fraud-proof). 
```

### 2. Security {#security}

To some extent, state channels rely on Ethereum to provide security and protect users from malicious peers. As discussed in later sections, channels use a fraud proof mechanism that lets users challenge attempts to finalize the channel with an invalid or stale update. 

In this case, the honest party provides the latest valid state of the channel as a fraud proof to the on-chain contract for verification. Fraud proofs enable mutually distrustful parties to conduct off-chain transactions without risking their funds in the process. 

### 3. Finality {#finality}

State updates collectively signed by channel users are considered as good as on-chain transactions. Still, all in-channel activity only achieves true finality when the channel is closed on Ethereum. 

In the optimistic case, both parties can cooperate and sign the final state update and submit on-chain to close the channel, after which the funds are distributed according to the channel's final state. In the pessimistic case, where someone tries to cheat by posting an incorrect state update on-chain, their transaction isn't finalized until the challenge window elapses.

> ## How do state channels work? {#how-do-state-channels-work}

Reviewer: It seems strange to me to have a section "How do state channels work" immediately after two chunky paragraphs that  already explain some aspects of how state channels work. Feels to me like this section needs a restructure so that "How state channels work" is a top level header, with several subheaders. "How do state channels interact with Ethereum" should be a subheader, for example. Please also try to minimize repetition of information.

Author: I applied the same structure used in the other scaling pages here: High-level overview of the protocol → description of its relationship with Ethereum → low-level explanation of the protocol's mechanics, etc. 

Maybe the chunkiness of preceding paragraphs is because I wanted to keep the original content (explanations of channels, payment channels, and state channels). I just edited the text to have the "How do state channels interact with Ethereum?" as an H4 header under the "How do state channels work?" section, though. If there are any other parts you want to see placed in the section, let me know.

Reviewer: yea I understand and appreciate what you've done - the level of detail is great and overall the page is improved. No doubt. My concerns are just a) whether the additional content makes the existing structure feel imbalanced and maybe disrupts the logic flow through the page, and b) the level of assumed prior knowledge required to parse some of the more complex sentences that arrive early. The original structure may have worked well with the original content but may need some refining now.

I don't think these need to be big structural changes necessarily - perhaps just making sure technical terms are defined on first use (even just by linking out to an external definition) and tweaking the headers would be enough (e.g. instead of "how does it work" we might change to "finer details" or "diving deeper" or something), just to help scaffold the reader a little more.

Author: Sure, the comment about structure makes sense. On linking to definitions of technical terms: I think @minimalsm commented about too many links in a page affecting reader experience (this was for the data availability page). 

Still, it's a good idea to provide a gentler on-ramp for beginner-level readers by including links to definitions. That's partly why I'm working on including some new definitions for the glossary--stuff like "Merkle proof" or "state transition" may sound like Greek to new readers, but having them in the glossary (and linking to it) can help.

Users in each virtual channel interact via a new contract instance, with the ledger channel able to support multiple contract instances. The ledger channel's state also contains more than one contract storage state, allowing for parallel execution of applications off-chain between different users.  

Just like regular channels, users exchange state updates to progress the state machine. Except a dispute arises, the intermediary only has to be contacted when opening or terminating the channel. 

### Virtual payment channels {#virtual-payment-channels}

Virtual payment channels work off the same idea as virtual state channels: participants connected to the same network can pass messages without needing to open a new channel on-chain. In virtual payment channels, value transfers are routed through one or more intermediaries, with guarantees that only the intended recipient can receive transferred funds. 

## Applications of state channels {#applications-of-state-channels}

> 

Reviewer: Might be nice to mention the potential for implementing a state-channel on a layer-2.

Author: I thought scaling solutions that don't use Ethereum for data availability don't fall under layer-2?

Reviewer: sure, but I remember discussions about whether a state-channel could be implemented posting its initial and final state to an L2 for additional scaling. 

e.g. Raiden on Arbitrum https://medium.com/raiden-network/the-raiden-network-now-live-on-arbitrum-3d57fedf7961

Reviewer: A simple one-liner like that is plenty in my opinion.

Author: Here's the final version I added:

> Implementing state channels on layer 2 solutions, such as [rollups](/developers/docs/scaling/#rollups), could make them even more attractive for payments. While channels offer cheap payments, the costs of setting up the on-chain contract on Mainnet during the opening phase can be get expensive—especially when gas fees spike. Ethereum-based rollups offer [lower transaction fees](https://l2fees.info/) and can reduce overhead for channel participants by bringing down setup fees.

Reviewer 2: nice

Other possible use-cases for state channel apps include ENS name ownership, NFT ledgers, and many more. 

### Atomic transfers {#atomic-transfers}

Early payment channels were restricted to transfers between two parties, limiting their usability. However, the introduction of virtual channels allowed individuals to route transfers through intermediaries (i.e., multiple p2p channels) without having to open a new channel on-chain. 

Commonly described as "multi-hop transfers", routed payments are atomic (i.e., either all parts of the transaction succeed or it fails altogether). Atomic transfers use [Hashed Timelock Contracts (HTLCs)](https://en.bitcoin.it/wiki/Hash_Time_Locked_Contracts) to ensure the payment is released only if certain conditions are met, thereby reducing counterparty risk. 

## Drawbacks of using state channels {#drawbacks-of-state-channels}

### Liveness assumptions {#liveness-assumptions}

To ensure efficiency, state channels place time limits on the ability of channel participants to respond to disputes. This rule assumes that peers will always be online to monitor channel activity and contest challenges when necessary. 

> In reality, users can go offline for reasons out of their control (e.g., poor Internet connection, mechanical failure, etc.). If an honest user goes offline, a malicious peer can exploit the situation by presenting old intermediate states to the adjudicator contract and stealing the committed funds. 

Reviewer 2: ```suggestion
In reality, users can go offline for reasons out of their control (e.g., poor internet connection, mechanical failure, etc.). If an honest user goes offline, a malicious peer can exploit the situation by presenting old intermediate states to the adjudicator contract and stealing the committed funds. 
```

- [Making Sense of Ethereum’s Layer 2 Scaling Solutions: State Channels, Plasma, and Truebit](https://medium.com/l4-media/making-sense-of-ethereums-layer-2-scaling-solutions-state-channels-plasma-and-truebit-22cb40dcc2f4) _– Josh Stark, Feb 12 2018_
- [State Channels - an explanation](https://www.jeffcoleman.ca/state-channels/) _Nov 6, 2015 - Jeff Coleman_
- [Basics of State Channels](https://education.district0x.io/general-topics/understanding-ethereum/basics-state-channels/) _District0x_
- [Blockchain State Channels: A State of the Art](https://ieeexplore.ieee.org/document/9627997)

**Payment channels**

> - [EthHub on payment channels](https://docs.ethhub.io/ethereum-roadmap/layer-2-scaling/payment-channels/)

Reviewer 3: ```suggestion
- [EthHub on payment channels](https://docs.ethhub.io/ethereum-roadmap/layer-2-scaling/payment-channels/)

_Know of a community resource that helped you? Edit this page and add it!_
```
